### DISCUSSION TITLE
Mixing FuncUI views and xaml views?

### DISCUSSION THREAD

**Discussion by Jamie-Rower** [JordanMarr/ReactiveElmish.Avalonia#28] [Q&A]
I am fairly new to F#, Avalonia, and Elmish, so forgive me my ignorance.  I really like the cleanliness of this solution, and in your readme you state that FuncUI can be used as desired, but I am experiencing some difficulty around this.

What I am attempting to do is incorporate a view created using FuncUI into a xaml view, similar to how you may use a component in Blazor/Angular/etc.  The ultimate goal of this is to attempt to register click and mouse move events on non-command elements, like a stack panel, and it seems as though I can add handlers using FuncUI.  The xaml interactivity options for this are not supported at the moment, or only for windows WPF which is not suitable for my project.

What I have attempted to do is to use:
let start view = 
    AvaloniaProgram.mkProgram init update bindings
    |> AvaloniaProgram.startElmishLoop view

with a custom defined view:
let view (model : Model) (dispatch : Msg -> unit) =
    StackPanel.create [
        StackPanel.children [
            TextBlock.create [
                TextBlock.text "Text"
            ]
        ]
    ]

Instead of using:
let vm : ElmishViewModel<Model,Msg> =
    AvaloniaProgram.mkSimple init update bindings
    |> ElmishViewModel.create
    |> ElmishViewModel.terminateOnViewUnloaded Terminate

as we are for the xaml views

This compiles, but I'm not sure if it works because we have things laid out in a tab control based on the model:
type Model = 
    {
        ContentVM: IElmishViewModel
    }

with tab content defined as
let update (msg: Msg) (model: Model) = 
    match msg with
    | ShowFuncUI ->
        { model with ContentVM = FuncViewModel.vm }
    | Terminate ->
        model

This doesn't work without the vm definition, obviously.  So the question is:

Is there a way to provide a custom view? Being able to '|> ElmishViewModel.create myView' for example.  I know that isn't currently possible, and I'm afraid my ignorance hides the difficulty of that task.

Any course correction, hints, or comments would be appreciated.

Thank you

**Comment by JordanMarr** [comment]
I haven't actually combined the two, but technically it should be possible.
I'll take a look at your example.

**Comment by JordanMarr** [comment]
I was able to create a component using FuncUI and splice it into the 'AvalloniaExample' project:

https://github.com/JordanMarr/Elmish.Avalonia/tree/FuncUI


Steps taken:
* Add the Avalonia.FuncUI package (obviously)
* Added `.UseSkia` to `AppBuilder` in Program.fs
* Added a "FuncUI" folder to hold the new component
* Added a new "Avalonia UserControl" named "CounterComponentUC.axaml"
* Added a new .fs file, "CounterComponentUC.fs
* Spliced it into my main view: `<local:CounterComponentUC DockPanel.Dock="Right"  />`

Note that the axaml file seems to be necessary, but it's Contents are replaced with the FuncUI view.

So you can definitely do it!

![image](https://github.com/JordanMarr/Elmish.Avalonia/assets/1030435/0575da3d-7c95-4eaf-8e40-fbf9615a48d9)

**Reply by Jamie-Rower** [reply]
Wow! Thank you so much!  I did not expect such a speedy and thorough response.  I was attempting something similar, but messier and more roundabout, and this is so much better; opens a lot of doors for me.  I'm actually going to try to reverse it, so the FuncUI view is the parent, but I'm pretty sure that should work too.  I really appreciate your response!
